### PR TITLE
Print numbers in a uniform way

### PR DESCRIPTION
**Pull request by lydell**
- Still preserve the radix (binary, octal, hexadecimal or decimal) used
  in the original source code.
- Still preserve scientific notation.
- Add 0 to fractions. `.1` -> `0.1`
- Remove trailing dots from integers. `1.` -> `1`, `1.e1` -> `1e1`
- Always print numbers lowercase. `0b`, `0o`, `0x`, `1e1`, `0xdeadbeef`
- Remove unneeded plus in scientific notation. `1e+1` -> `1e1`
- Remove unneeded zeroes in scientific notation. `1e-001` -> `1e-1`
- Preserve leading zeroes in non-decimal radix. This can be useful when
  working in binary, and having both `0b111000` and `0b000111` for
  example.

### CODE REVIEW

**Comment by lydell** [comment]
`jest -u` failed because of this flow bug: facebook/flow#3283

File: src/printer.js [hunk 1]
```diff
@@ -2537,6 +2537,20 @@ function makeString(rawContent, enclosingQuote) {
...
+    .replace(/[a-f]/g, match => match.toUpperCase())
+    // Lowercase radix letter.
+    .replace(/^0[BOX]/, match => match.toLowerCase())
+    // Lowercase scientific notation, with unnecessary plus and zeroes removed.
+    .replace(/^([\d.]+)[eE](?:\+|(-))?0*/, "$1e$2")
+    // Make sure numbers always start with a digit.
+    .replace(/^\./, "0.")
+    // Remove trailing dot.
+    .replace(/\.$/, "");
+}
```
Comment (lydell) [inline src/printer.js]: Here is prettier's formatting of this block:

```js
  return rawNumber.// Uppercase hex digits.
  replace(/[a-f]/g, match => match.toUpperCase()).// Lowercase radix letter.
  replace(
    /^0[BOX]/,
    match => match.toLowerCase()
  ).// Lowercase scientific notation, with unnecessary plus and zeroes removed.
  replace(
    /^([\d.]+)[eE](?:\+|(-))?0*/,
    "$1e$2"
  ).// Make sure numbers always start with a digit.
  replace(/^\./, "0.").// Remove trailing dot.
  replace(/\.$/, "");
```

Comment (vjeux) [inline src/printer.js]: There's already an issue for that FYI. I'm hoping to address it with #469

**Comment by vjeux** [comment]
> Always print hexadecimal digits uppercase. 0x123ABCDEF

I searched for 0xdeadbeef which is a hex number we can grep for and for JavaScript, the majority of code is written lowercase on the 20+ pages I browsed.
https://github.com/search?l=JavaScript&p=1&q=0xdeadbeef&type=Code&utf8=%E2%9C%93

Also the Chrome element inspector displays them lowercase

image

**Comment by vjeux** [comment]
> This can be useful when working in binary, and having both 0b111000 and 0b000111 for example.

TIL that JavaScript had support for binary prefix!

**Comment by vjeux** [comment]
It feels overkill to me since those variants are very rare in practice, but I'm not opposed to it. You can comment the example that breaks on the flow parser so that we can ship it :)

**Comment by hzoo** [comment]
> TIL that JavaScript had support for binary prefix!

It's an ES6 feature so not as much use probably 😄? http://babeljs.io/docs/plugins/transform-es2015-literals/

**Comment by lydell** [comment]
I’ve pushed a commit to always print numbers lowercase.

What should we do about the tests and flow?

**Comment by vjeux** [comment]
I think we should just run the test on babylon. Since the bug doesn't affect the canonical output, we are not going to generate bad code for flow.

**Comment by vjeux** [comment]
I'll also let @jlongster review this one to make sure he is onboard with this.

**Comment by jlongster** [comment]
I think this is a great use case for enforcing consistent style. Thanks!

My only question is the code doesn't seem to implement some of the things you mentioned, like "Preserve leading zeroes in non-decimal radix" ? Was it doing that before and you were just listing all of the behaviors you want? I think `node.raw` should have done that.

**Comment by lydell** [comment]
> My only question is the code doesn't seem to implement some of the things you mentioned, like "Preserve leading zeroes in non-decimal radix" ? Was it doing that before and you were just listing all of the behaviors you want?

All of this is correct. This applies to the two bullet points starting with "Still".

**Comment by vjeux** [comment]
@lydell if you send an update in the next hour, this will be part of the next release :)

**Comment by lydell** [comment]
Sorry, what would you like to be updated? :)

**Comment by vjeux** [comment]
Travis is failing :) I think you just need to change the parser in jsfmt.spec.js to be babylon and that should fix it.

**Comment by vjeux** [comment]
Also need to re-record the tests: 

image

**Comment by lydell** [comment]
Switched to babylon and updated snapshots.

**Comment by vjeux** [comment]
yay, thanks!

**Pull request merged**